universal/src/compression/zstd.cpp

```diff
@@ -0,0 +1,95 @@
+#include <userver/compression/zstd.hpp>
+
+#include <zstd.h>
+#include <zstd_errors.h>
+
+USERVER_NAMESPACE_BEGIN
+
+namespace compression::zstd {
+
+namespace {
+// The same size as in ZSTD_DStreamOutSize();
+const size_t kDecompressBufferSize = ZSTD_DStreamOutSize();
+}  // namespace
+
+std::string DecompressStream(std::string_view compressed, size_t max_size) {
+  std::string decompressed;
+  std::string buf(kDecompressBufferSize, '\0');
+
+  auto* stream = ZSTD_createDStream();
+  if (stream == nullptr) {
+    throw std::runtime_error("Couldn't create ZSTD decompression stream");
+  }
+
+  {
+    auto err_code = ZSTD_DCtx_reset(
+        stream, ZSTD_ResetDirective::ZSTD_reset_session_and_parameters);
+    if (ZSTD_isError(err_code)) {
+      ZSTD_freeDStream(stream);
+      throw ErrWithCode(ZSTD_getErrorName(err_code));
+    }
+  }
+
+  auto stream_del = [](ZSTD_DStream* ptr) { ZSTD_freeDStream(ptr); };
+  auto stream_guard =
+      std::unique_ptr<ZSTD_DStream, decltype(stream_del)>(stream, stream_del);
+
+  for (size_t cur_pos(0); cur_pos < compressed.size();) {
+    ZSTD_inBuffer input{
+        compressed.data() + cur_pos,
+        std::min(kDecompressBufferSize, compressed.size() - cur_pos), 0};
+    ZSTD_outBuffer output{buf.data(), buf.size(), 0};
+    auto* output_pos = static_cast<char*>(output.dst);
+
+    while (input.pos < input.size) {
+      const auto ret = ZSTD_decompressStream(stream, &output, &input);
+
+      if (ZSTD_isError(ret)) {
+        throw ErrWithCode(ZSTD_getErrorName(ret));
+      }
+
+      decompressed.append(output_pos, output_pos + output.pos);
+      output_pos = static_cast<char*>(output.dst) + output.pos;
+    }
+
+    if (decompressed.size() > max_size) {
+      throw TooBigError();
+    }
+
+    cur_pos += input.size;
+  }
+
+  return decompressed;
+}
+
+std::string Decompress(std::string_view compressed, size_t max_size) {
+  auto decompressed_size =
+      ZSTD_getFrameContentSize(compressed.data(), compressed.size());
+
+  switch (decompressed_size) {
+    case ZSTD_CONTENTSIZE_UNKNOWN:
+      return DecompressStream(compressed, max_size);
+    case ZSTD_CONTENTSIZE_ERROR:
+      throw std::runtime_error("Error while getting size");
+    default:
+      if (decompressed_size > max_size) {
+        throw TooBigError();
+      }
+      if (decompressed_size > std::numeric_limits<int64_t>::max() / 2) {
+        return DecompressStream(compressed, max_size);
```

Comment:
Are you sure you want to decompress more than 4611686018427387903 bytes?
seems like a sign mismatch here for `<>`
also decompressed_size is unsigned long long